### PR TITLE
add paged-attention

### PR DESCRIPTION
**Pull request by kaiyuanm**
1. Add paged attention operation commonly used in large language model inference. About paged attention view https://arxiv.org/pdf/2309.06180.pdf.
2. This implementation requires triton 2.2.0.

There is some performance data between triton and vllm.
```
vllm_paged_attention-B32-G8-D128-bs16-v2:
   context_len     triton  vllm-0.3.0
0        512.0   6.618643    3.582180
1       1024.0   7.798634    3.925020
2       2048.0   8.560921    4.286430
3       4096.0  12.241113    4.460762
4       8192.0  13.589705    4.558541
5      16384.0  14.094599    4.609148
```
Note:
- Input layout is different, performance is for reference only.
- Some input shapes is still being optimized.

### CODE REVIEW

File: src/flag_attn/paged.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+            block_tables_ptr + seq_idx * max_num_blocks_per_seq + block_idx
+        )
+
+        # Load a key block.
+        kv_block_offset = block_number * stride_kv0 + kv_offset
+        mask_offset = block_idx * KV_BLOCK_SIZE + block_offset
+        kv_mask = mask_offset[:, None] < context_len
+
+        # k: [KV_BLOCK_SIZE, HEAD_SIZE]
+        k = tl.load(k_cache_ptr + kv_block_offset, mask=kv_mask, other=0.0)
```
Comment (tongxin) [inline src/flag_attn/paged.py:272]: Is it possible to omit kv_mask here and the mask the output?

File: src/flag_attn/paged.py [hunk 2]
```diff
@@ -0,0 +1,415 @@
...
+        if PADDED_QUERY_GROUP_SIZE == 1:
+            qk = tl.sum(q[:, None, :] * k[None, :, :], axis=2)
+        else:
+            qk = tl.dot(q, k.T, out_dtype=tl.float32)
+
+        qk *= attn_scale
+        qk = tl.where(mask_offset < context_len, qk, float("-inf"))
+
+        # m_ij: [PADDED_QUERY_GROUP_SIZE]
+        m_ij = tl.max(qk, axis=1)
```
Comment (tongxin) [inline src/flag_attn/paged.py:284]: What does subscript j indicate?

File: src/flag_attn/paged.py [hunk 3]
```diff
@@ -0,0 +1,415 @@
...
+        m_i_new = tl.maximum(m_i, m_ij)
+
+        # p: [PADDED_QUERY_GROUP_SIZE, KV_BLOCK_SIZE]
+        p = tl.exp(qk - m_i_new[:, None])
+        alpha = tl.exp(m_i - m_i_new)
```
Comment (tongxin) [inline src/flag_attn/paged.py:289]: Could we use tl.math.exp2 for better performance and consistency?

File: examples/paged_example.py [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    torch.set_default_dtype(dtype)
+    torch.set_default_device(device=device)
+
+    torch.cuda.manual_seed(seed)
+
+    num_query_heads, num_kv_heads = num_heads
+
+    context_lens = torch.randint(1, MAX_SEQ_LEN, [num_seqs], dtype=torch.int32)
+    max_context_len = context_lens.max().item()
+    max_num_blocks_per_seq = (max_context_len + block_size - 1) // block_size
```
Comment (iclementine) [inline examples/paged_example.py:28]: In autogressive decoding, does the size og block_table gradully grow to 
`(num_seqs, ceil_div(MAX_SEQ_LEN, NUM_BLOCKS))`

File: src/flag_attn/paged.py [hunk 5]
```diff
@@ -0,0 +1,415 @@
...
+    max_num_blocks_per_seq = block_tables.shape[1]
+
+    if query_group_size == 1:
+        padded_group_size = 1
+    elif query_group_size < 16:
+        padded_group_size = 16
+    else:
+        padded_group_size = triton.next_power_of_2(query_group_size)
+
+    assert head_size in [64, 128, 256], f"head_size={head_size}"
```
[the remark this answers is not in the document]
Comment (iclementine) [reply src/flag_attn/paged.py:33]: what about adding 32 as acceptable head dim?

File: src/flag_attn/paged.py [hunk 6]
```diff
@@ -0,0 +1,415 @@
...
+    assert query_group_size in [1, 2, 4, 8, 16, 32, 64, 128, 256]
+
+    # config for A100
+    device = torch.cuda.device_of(query)
+    num_sms = torch.cuda.get_device_properties(device).multi_processor_count
+    if num_splits == 0:
+        if num_seqs * num_kv_heads > 2 * num_sms:
+            num_splits = 1
+            if max_context_len >= 4096:
+                partition_size = max(256, kv_block_size)
```
Comment (iclementine) [inline src/flag_attn/paged.py:44]: Should `partition_size` always be a multiple of `kv_block_size`?

File: src/flag_attn/paged.py [hunk 7]
```diff
@@ -0,0 +1,415 @@
...
+        partition_size = triton.cdiv(max_context_len, num_splits)
+        partition_size = triton.next_power_of_2(partition_size)
+        assert partition_size >= kv_block_size
+
+    with torch.cuda.device(device):
+        if num_splits == 1:
+            grid = (num_seqs, num_kv_heads, 1)
+            _paged_attn_kernel[grid](
+                out,
+                out,
```
Comment (iclementine) [inline src/flag_attn/paged.py:61]: It is better to mark the first 2 inputs as dummy inputs to indicate that they are not needed in this case.

File: src/flag_attn/paged.py [hunk 8]
```diff
@@ -0,0 +1,415 @@
...
+    kv_head_idx = tl.program_id(1)
+
+    context_len = tl.load(context_lens_ptr + seq_idx)
+
+    num_partitions = tl.cdiv(context_len, PARTITION_SIZE)
+    group_head_offset = (
+        tl.arange(0, QUERY_GROUP_SIZE)[:, None] * HEAD_SIZE
+        + tl.arange(0, HEAD_SIZE)[None, :]
+    )
+    if num_partitions == 1:
```
Comment (iclementine) [inline src/flag_attn/paged.py:359]: Is this a dummy case(when no partitioning is used, this kernel is not launched)?

File: benchmark/paged_benchmark.py [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+    context_lens = torch.zeros(num_seqs, dtype=torch.int32, device=device) + context_len
+    max_num_blocks_per_seq = (context_len + block_size - 1) // block_size
+
+    attn_scale = head_size**-0.5
+    q = torch.empty(num_seqs, num_query_heads, head_size, dtype=dtype, device=device)
+    q.uniform_(-attn_scale, attn_scale)
+    out = torch.empty_like(q)
+
+    k_cache = torch.empty(
+        NUM_BLOCKS, num_kv_heads, block_size, head_size, dtype=dtype, device=device
```
[the remark this answers is not in the document]
Comment (iclementine) [reply benchmark/paged_benchmark.py:139]: There is no standard layout of kv cache, but I think the best layout is the one that considers operators before and after an attention operator with a transformer layer. Thus，`Shape(batch_size, seqlen_kv, num_heads_kv, head_dim)` and row-major is the best.

But the kernel taks as inputs the stride of some input tensors explicitly, so it handles inputs with different layouts.

File: benchmark/paged_benchmark.py [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+            block_tables,
+            context_lens,
+            block_size,
+            context_len,
+            PARTITION_SIZE=512,
+            version=vllm_version,
+        )
+        ms = triton.testing.do_bench(fn, warmup=warmup, rep=rep)
+
+    total_flops = 2.0 * num_seqs * num_query_heads * 2 * context_len * head_size
```
[the remark this answers is not in the document]
Comment (kaiyuanm) [reply benchmark/paged_benchmark.py:184]: S=Q·K.T and SV need to be calculated twice

File: src/flag_attn/paged.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (kaiyuanm) [reply src/flag_attn/paged.py:33]: ok

File: src/flag_attn/paged.py [hunk 11]
```diff
@@ -0,0 +1,415 @@
...
+                num_splits = triton.cdiv(max_context_len, partition_size)
+        else:
+            partition_size = max(256, kv_block_size)
+            num_splits = triton.cdiv(max_context_len, partition_size)
+            if max_context_len <= 1024 or kv_block_size >= 256:
+                num_splits = 1
+    elif num_splits > 1:
+        partition_size = triton.cdiv(max_context_len, num_splits)
+        partition_size = triton.next_power_of_2(partition_size)
+        assert partition_size >= kv_block_size
```
[the remark this answers is not in the document]
Comment (kaiyuanm) [reply src/flag_attn/paged.py:54]: Refer to flash_attn flash_attn_with_kvcache definition.
> num_splits: int. If > 1, split the key/value into this many chunks along the sequence.
           If num_splits == 1, we don't split the key/value. If num_splits == 0, we use a heuristic
           to automatically determine the number of splits.

File: src/flag_attn/paged.py (hunk 3 above)
Comment (kaiyuanm) [reply src/flag_attn/paged.py:289]: I'll try it.

File: src/flag_attn/paged.py (hunk 7 above)
Comment (kaiyuanm) [reply src/flag_attn/paged.py:61]: ok

File: src/flag_attn/paged.py (hunk 6 above)
Comment (kaiyuanm) [reply src/flag_attn/paged.py:44]: For now, yes.

File: examples/paged_example.py (hunk 4 above)
Comment (kaiyuanm) [reply examples/paged_example.py:28]: I think so, this block_table is just for this test case.

File: src/flag_attn/paged.py (hunk 2 above)
Comment (kaiyuanm) [reply src/flag_attn/paged.py:284]: Here j has nothing to do with the dimension, it is just a mark indicating this iteration, I'll fix it.

File: src/flag_attn/paged.py (hunk 1 above)
Comment (kaiyuanm) [reply src/flag_attn/paged.py:272]: Looks like ok. What do you think? @iclementine

File: benchmark/paged_benchmark.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (tongxin) [reply benchmark/paged_benchmark.py:184]: I got it, thanks.

File: src/flag_attn/paged.py (hunk 1 above)
Comment (iclementine) [reply src/flag_attn/paged.py:292]: Wehen loading, the mask is obligatory if it would access illegal memory otherwise. However you can use the modulo trick to avoid masking. Whether it is more efficient than masking depends on the cost .

**Review by iclementine — APPROVED** [review]
lgtm
